### ISSUE TITLE
Rendering error in Mapzen tile 9/123/184.topojson

### ISSUE THREAD

**Issue by mbloch** [tangrams/tangram#456]
URL of map in Tangram Play showing the issue:
https://mapzen.com/tangram/play/?lines=5&scene=https%3A%2F%2Fapi.github.com%2Fgists%2F4b13717189152290d4b89f52f44d335a#9.9159/44.8688/-93.1499
URL of affected tile:  https://tile.mapzen.com/mapzen/vector/v1/all/9/123/184.topojson

You should be able to see in the example map above and the images below that one particular polygon in the water layer is drawing incorrectly. If I switch to an MVT or GeoJSON datasource, the problem goes away.

![image](https://cloud.githubusercontent.com/assets/1232222/20635946/c4c8e49c-b332-11e6-9867-34c4cba8c850.png)
![image](https://cloud.githubusercontent.com/assets/1232222/20636152/b80a8a64-b335-11e6-8112-e0d1c5d4a34c.png)

If I open the TopoJSON tile in Mapshaper or another viewer, the affected polygon looks normal, so the oddly rendered shape isn't baked into the tile data:

![image](https://cloud.githubusercontent.com/assets/1232222/20635941/a540b032-b332-11e6-9331-1daebf16a370.png)

**Comment by mbloch** [comment]
Here's another example of the same issue, also in the water layer:
![image](https://cloud.githubusercontent.com/assets/1232222/20819222/fdb4be48-b800-11e6-9f04-528828cd01c5.png)

Link to the above map view in Tangram Play (Cinnabar theme, TopoJSON tile source):
https://mapzen.com/tangram/play/?scene=https%3A%2F%2Fmapzen.com%2Fcarto%2Fcinnabar-style%2Fcinnabar-style.yaml#9.6327/37.2879/-77.1609

**Comment by bcamper** [comment]
Thanks for the reports, we'll take a look soon!

**Comment by bcamper** [comment]
Looked at this with @rmarianski. These water polygons are quite detailed and there are small differences in the encoding between all 3 formats. We haven't found the problem yet, but my suspicion is that with these very small and complex polygons (in the first example, it is a polygon with an outer ring of 78 points, and 21 inner rings of ~5 points each), if there is any loss of precision (or small transform errors), perhaps some of the rings are ending up coincident or intersecting. This is the kind of thing that could cause a tessellator like the earcut library to freak out, but might not show up in all viewers.

@rmarianski can you help investigate further on the tiles side?

**Comment by rmarianski** [comment]
I loaded up the topojson in qgis, and it looks fine there too. I'll try to drill down a little bit more into the details of the geometry to see if we can find anything.

**Comment by bcamper** [comment]
I'm wondering if any rings may be touching in a way that qgis is cool with
but other more aggressive tessellators are not?

On Tue, Dec 13, 2016 at 10:07 AM, Robert Marianski <notifications@github.com

**Comment by rmarianski** [comment]
There are ogc invalid geometries in that tile, so it's likely something like that is going on.

**Comment by rmarianski** [comment]
We increased the scale used on the topojson tiles, which looks to have fixed the 2 examples you noticed here.

**Comment by rmarianski** [comment]
Well, I thought it did, but I must have been looking at a cached version. When I force reload it looks broken to me again unfortunately.

**Comment by rmarianski** [comment]
And indeed it was a cache issue, but now it should look ok.

**Comment by mbloch** [comment]
Thanks for looking into this issue :)

Increasing the TopoJSON scale is a simple fix (although perhaps not 100% reliable).

FWIW, I spent a bit of time trying to find a simple example that would display the problem. I made a TopoJSON tile containing a polygon with a single hole that triggers the tesselation error. (The shapes were extracted from the first example I posted above).

 I'm thinking I should add an option to Mapshaper that prevents the kind of topology error that you see in this example....

Here's a link to the example on Mapzen Play:
https://mapzen.com/tangram/play/?scene=https%3A%2F%2Fgist.githubusercontent.com%2Fanonymous%2F8f10372a22663c2679bdcec5d5209db8%2Fraw%2F4bbc13851524dd3e3792dda479ae15a41e067649%2Fscene.yaml#11.1856/44.9273/-93.1748

Here's a link to the file:
http://int.nyt.com/newsgraphics/map-tiles/2016/precincts/debug/v9/9/123/184.topojson

Here's a screengrab showing the problem area:
![image](https://cloud.githubusercontent.com/assets/1232222/21275943/29cee664-c39d-11e6-8fdb-1e70e0283d63.png)
